File: ballast-core/npm/src/domain/models/tile.ts
import { CubicCoordinates } from './cubic-coordinates';
import { Terrain } from './terrain';
import { TileShape } from './tile-shape';

export class Tile {

    public readonly cubicCoordinates: CubicCoordinates;
    public readonly tileShape: TileShape;
    private _terrain: Terrain;

    public get terrain(): Terrain {
        return this._terrain;
    }

    public set terrain(value: Terrain) { 
        this.setTerrain(value);
    }
    
    public constructor(cubicCoordinates: CubicCoordinates, tileShape: TileShape, terrain: Terrain) {
        this.cubicCoordinates = cubicCoordinates;
        this.tileShape = tileShape;
        this._terrain = terrain;
    }

    public setTerrain(terrain: Terrain) {
        this._terrain = terrain || this.terrain;
    }

}

File: ballast-core/dotnet/src/domain/models/Tile.cs
namespace Ballast.Core.Domain.Models
{
    public class Tile
    {

        public CubicCoordinates CubicCoordinates { get; private set; }
        public TileShape TileShape { get; private set; }
        public Terrain Terrain { get; private set; }

        public Tile(CubicCoordinates cubicCoordinates, TileShape tileShape, Terrain terrain)
        {
            CubicCoordinates = cubicCoordinates;
            TileShape = tileShape;
            Terrain = terrain;
        }

        public void SetTerrain(Terrain terrain)
        {
            this.Terrain = terrain;
        }

    }
}